vm/src/exceptions.rs

```diff
@@ -195,11 +195,16 @@ pub fn write_exception<W: Write>(
     exc: &PyBaseExceptionRef,
 ) -> Result<(), W::Error> {
     if let Some(cause) = exc.cause() {
-        write_exception(output, vm, &cause)?;
-        writeln!(
-            output,
-            "\nThe above exception was the direct cause of the following exception:\n"
-        )?;
+        // This can be a special case: `raise e from e`,
+        // we just ignore it and treat like `raise e`.
+        // So, doing nothing here.
+        if !cause.as_object().is(&exc) {
```

Comment:
Though this handles `raise e from e` it still segfaults in other situations:

```
class E(Exception): pass
class D(E): pass

e, d = E(), D()
raise d from e
```

CPython uses a set to help stop the infinite recursion, constantly filling it up with exceptions until it sees them all. Something similar can be done here (a new private function that `write_exception` calls seems better as to not require callers of this public function to supply the set).

Reply:
Good catch! I've refactored this into a `write_exception_recursive` and added several new tests.
Please, take a look 🙂